---
title: "Execution commands"
author: "Georgios A. Manios"
 
output: html_document
---

 
 
# In this part, we show how each tool was executed with the following commands:

## PRED-LD
```{text}
### PRED-LD imputation tasks (Combined panels)
time python pred_ld.py --file-path ADHD.txt --r2threshold 0.5 --pop EUR --maf 0 --ref all_panels 
time python pred_ld.py --file-path UACR.txt --r2threshold 0.5 --pop EUR --maf 0 --ref all_panels 
time python pred_ld.py --file-path GFREUR.txt --r2threshold 0.5 --pop EUR --maf 0 --ref all_panels 
time python pred_ld.py --file-path EP.txt --r2threshold 0.5 --pop EUR --maf 0 --ref all_panels 
time python pred_ld.py --file-path CAD.txt --r2threshold 0.5 --pop EUR --maf 0 --ref all_panels 
time python pred_ld.py --file-path DLID.txt --r2threshold 0.5 --pop EAS --maf 0 --ref all_panels 
time python pred_ld.py --file-path CCANCER.txt --r2threshold 0.5 --pop EAS --maf 0 --ref all_panels 
time python pred_ld.py --file-path GFRAFR.txt --r2threshold 0.5 --pop AFR --maf 0 --ref all_panels 


### PRED-LD imputation tasks (TOP-LD)
time python pred_ld.py --file-path ADHD.txt --r2threshold 0.5 --pop EUR --maf 0 --ref TOP_LD
time python pred_ld.py --file-path UACR.txt --r2threshold 0.5 --pop EUR --maf 0 --ref TOP_LD 
time python pred_ld.py --file-path GFREUR.txt --r2threshold 0.5 --pop EUR --maf 0 --ref TOP_LD 
time python pred_ld.py --file-path EP.txt --r2threshold 0.5 --pop EUR --maf 0 --ref TOP_LD 
time python pred_ld.py --file-path CAD.txt --r2threshold 0.5 --pop EUR --maf 0 --ref TOP_LD 
time python pred_ld.py --file-path DLID.txt --r2threshold 0.5 --pop EAS --maf 0 --ref TOP_LD
time python pred_ld.py --file-path CCANCER.txt --r2threshold 0.5 --pop EAS --maf 0 --ref TOP_LD 
time python pred_ld.py --file-path GFRAFR.txt --r2threshold 0.5 --pop AFR --maf 0 --ref TOP_LD 


### PRED-LD imputation tasks (Pheno Scanner)
time python pred_ld.py --file-path ADHD.txt --r2threshold 0.5 --pop EUR --maf 0 --ref Pheno_Scanner
time python pred_ld.py --file-path UACR.txt --r2threshold 0.5 --pop EUR --maf 0 --ref Pheno_Scanner
time python pred_ld.py --file-path GFREUR.txt --r2threshold 0.5 --pop EUR --maf 0 --ref Pheno_Scanner 
time python pred_ld.py --file-path EP.txt --r2threshold 0.5 --pop EUR --maf 0 --ref Pheno_Scanner
time python pred_ld.py --file-path CAD.txt --r2threshold 0.5 --pop EUR --maf 0 --ref Pheno_Scanner 
time python pred_ld.py --file-path DLID.txt --r2threshold 0.5 --pop EAS --maf 0 --ref Pheno_Scanner
time python pred_ld.py --file-path CCANCER.txt --r2threshold 0.5 --pop EAS --maf 0 --ref Pheno_Scanner
time python pred_ld.py --file-path GFRAFR.txt --r2threshold 0.5 --pop AFR --maf 0 --ref Pheno_Scanner 


### PRED-LD imputation tasks (HapMap)
time python pred_ld.py --file-path ADHD.txt --r2threshold 0.5 --pop TSI --maf 0 --ref Hap_Map
time python pred_ld.py --file-path ADHD.txt --r2threshold 0.5 --pop CEU --maf 0 --ref Hap_Map

time python pred_ld.py --file-path UACR.txt --r2threshold 0.5 --pop TSI --maf 0 --ref Hap_Map
time python pred_ld.py --file-path UACR.txt --r2threshold 0.5 --pop CEU --maf 0 --ref Hap_Map

time python pred_ld.py --file-path GFREUR.txt --r2threshold 0.5 --pop TSI --maf 0 --ref Hap_Map
time python pred_ld.py --file-path GFREUR.txt --r2threshold 0.5 --pop CEU --maf 0 --ref Hap_Map

time python pred_ld.py --file-path EP.txt --r2threshold 0.5 --pop TSI --maf 0 --ref Hap_Map
time python pred_ld.py --file-path EP.txt --r2threshold 0.5 --pop CEU --maf 0 --ref Hap_Map

time python pred_ld.py --file-path CAD.txt --r2threshold 0.5 --pop TSI --maf 0 --ref Hap_Map
time python pred_ld.py --file-path CAD.txt --r2threshold 0.5 --pop CEU --maf 0 --ref Hap_Map


time python pred_ld.py --file-path DLID.txt --r2threshold 0.5 --pop CHD --maf 0 --ref Hap_Map
time python pred_ld.py --file-path DLID.txt --r2threshold 0.5 --pop CHB --maf 0 --ref Hap_Map
time python pred_ld.py --file-path DLID.txt --r2threshold 0.5 --pop JPT --maf 0 --ref Hap_Map


time python pred_ld.py --file-path CCANCER.txt --r2threshold 0.5 --pop CHD --maf 0 --ref Hap_Map
time python pred_ld.py --file-path CCANCER.txt --r2threshold 0.5 --pop CHB --maf 0 --ref Hap_Map
time python pred_ld.py --file-path CCANCER.txt --r2threshold 0.5 --pop JPT --maf 0 --ref Hap_Map

time python pred_ld.py --file-path GFRAFR.txt --r2threshold 0.5 --pop MKK --maf 0 --ref Hap_Map 
time python pred_ld.py --file-path GFRAFR.txt --r2threshold 0.5 --pop LWK --maf 0 --ref Hap_Map 
time python pred_ld.py --file-path GFRAFR.txt --r2threshold 0.5 --pop YRI --maf 0 --ref Hap_Map 
time python pred_ld.py --file-path GFRAFR.txt --r2threshold 0.5 --pop ASW --maf 0 --ref Hap_Map 
```
 
 
##   RAISS
 
### Performance grid search, masking  ~1/10 of the number of the intial SNPs
```{text}
raiss --ld-folder ld/EUR/ --ref-folder ref/EUR/ --gwas ADHD --chrom chr1 --ld-type plink performance-grid-search --harmonized-folder harm --masked-folder mask --imputed-folder imp --eigen-ratio-grid '[0.000001, 0.001,0.1]' --ld-threshold-grid '[0,4,10,20]' --output-path out 
--N-to-mask 1516


raiss --ld-folder ld/EUR/ --ref-folder ref/EUR/ --gwas UACR --chrom chr1 --ld-type plink performance-grid-search --harmonized-folder harm --masked-folder mask --imputed-folder imp --eigen-ratio-grid '[0.000001, 0.001,0.1]' --ld-threshold-grid '[0,4,10,20]'  --output-path out 
--N-to-mask 727


raiss --ld-folder ld/EUR/ --ref-folder ref/EUR/ --gwas GFREUR --chrom chr1 --ld-type plink performance-grid-search --harmonized-folder harm --masked-folder mask --imputed-folder imp --eigen-ratio-grid '[0.000001, 0.001,0.1]' --ld-threshold-grid '[0,4,10,20]' --output-path out --N-to-mask 508


raiss --ld-folder ld/EUR/ --ref-folder ref/EUR/ --gwas EP --chrom chr1 --ld-type plink performance-grid-search --harmonized-folder harm --masked-folder mask --imputed-folder imp --eigen-ratio-grid '[0.000001, 0.001,0.1]' --ld-threshold-grid '[0,4,10,20]'  --output-path out 
--N-to-mask 29503


raiss --ld-folder ld/EUR/ --ref-folder ref/EUR/ --gwas CAD --chrom chr1 --ld-type plink performance-grid-search --harmonized-folder harm --masked-folder mask --imputed-folder imp --eigen-ratio-grid '[0.000001, 0.001,0.1]' --ld-threshold-grid '[0,4,10,20]'  --output-path out --N-to-mask 55713


raiss --ld-folder ld/EAS/ --ref-folder ref/EAS/ --gwas DLID --chrom chr1 --ld-type plink performance-grid-search --harmonized-folder harm --masked-folder mask --imputed-folder imp --eigen-ratio-grid '[0.000001, 0.001,0.1]' --ld-threshold-grid '[0,4,10,20]'  --output-path out --N-to-mask 3127


raiss --ld-folder ld/EAS/ --ref-folder ref/EAS/ --gwas CCANCER --chrom chr1 --ld-type plink performance-grid-search --harmonized-folder harm --masked-folder mask --imputed-folder imp --eigen-ratio-grid '[0.000001, 0.001,0.1]' --ld-threshold-grid '[0,4,10,20]'  --output-path out --N-to-mask 6882


raiss --ld-folder ld/AFR/ --ref-folder ref/AFR/ --gwas GFRAFR --chrom chr1 --ld-type plink performance-grid-search --harmonized-folder harm --masked-folder mask --imputed-folder imp --eigen-ratio-grid '[0.000001, 0.001,0.1]' --ld-threshold-grid '[0,4,10,20]'  --output-path out --N-to-mask 463
``` 

###  Imputation tasks 

After performing performance grid search to find optimal hyper-parameters, we ended up to the below eigen-thresholds and minimum-ld values for each gwas study, as these parameters gave the best trade-off of lowest imputation error and highest fraction of masked SNPs.  
<br>
 
Reference panel: 1000 Genomes Project Phase 3, hg38 (as .bim file)   https://www.cog-genomics.org/plink/2.0/resources
 <br>
 

 
 
Precomputed LD matrices : https://zenodo.org/records/10391372

```{text}
time raiss --chrom chr1 --gwas ADHD --ld-folder ./ld/EUR --ref-folder ref/EUR/ --zscore-folder ./ --output-folder out/ --eigen-threshold 0.1 --minimum-ld 0 --R2-threshold 0.5

time raiss --chrom chr1 --gwas UACR --ld-folder ./ld/EUR --ref-folder ref/EUR/ --zscore-folder ./ --output-folder out/ --eigen-threshold 0.1 --minimum-ld 0 --R2-threshold 0.5

time raiss --chrom chr1 --gwas GFREUR --ld-folder ./ld/EUR --ref-folder ref/EUR/ --zscore-folder ./ --output-folder out/ --eigen-threshold 0.1 --minimum-ld 0 --R2-threshold 0.5

time raiss --chrom chr1 --gwas EP --ld-folder ./ld/EUR --ref-folder ref/EUR/ --zscore-folder ./ --output-folder out/ --eigen-threshold 0.1 --minimum-ld 0 --R2-threshold 0.5

time raiss --chrom chr1 --gwas CAD --ld-folder ./ld/EUR --ref-folder ref/EUR/ --zscore-folder ./ --output-folder out/ --eigen-threshold 0.1 --minimum-ld 4 --R2-threshold 0.5

time raiss --chrom chr1 --gwas DLID --ld-folder ./ld/EAS --ref-folder ref/EAS/ --zscore-folder ./ --output-folder out/ --eigen-threshold 0.1 --minimum-ld 0 --R2-threshold 0.5

time raiss --chrom chr1 --gwas CCANCER --ld-folder ./ld/EAS --ref-folder ref/EAS/ --zscore-folder ./ --output-folder out/ --eigen-threshold 0.1 --minimum-ld 0 --R2-threshold 0.5

time raiss --chrom chr1 --gwas GFRAFR --ld-folder ./ld/AFR --ref-folder ref/AFR/ --zscore-folder ./ --output-folder out/ --eigen-threshold 0.000001 --minimum-ld 0 --R2-threshold 0.5
``` 

 
##		SSIMP
Reference panel: 1000 Genomes Project Phase 3 (hg19) https://hgdownload.cse.ucsc.edu/gbdb/hg19/1000Genomes/phase3/
```{text}
time ssimp --gwas ADHD.txt ~/reference_panels/1000genomes/ALL.chr1.phase3_shapeit2_mvncall_integrated_v5a.20130502.genotypes.vcf.gz --out SSIMP_ADHD_EUR_chr1.txt --sample.names ~/reference_panels/1KG/integrated_call_samples_v3.20130502.ALL.panel/sample/super_pop=EUR

time ssimp --gwas UACR.txt ~/reference_panels/1000genomes/ALL.chr1.phase3_shapeit2_mvncall_integrated_v5a.20130502.genotypes.vcf.gz --out SSIMP_UACR_EUR_chr1.txt --sample.names ~/reference_panels/1KG/integrated_call_samples_v3.20130502.ALL.panel/sample/super_pop=EUR

time ssimp --gwas GFREUR.txt ~/reference_panels/1000genomes/ALL.chr1.phase3_shapeit2_mvncall_integrated_v5a.20130502.genotypes.vcf.gz --out SSIMP_GFR_EUR_chr1.txt --sample.names ~/reference_panels/1KG/integrated_call_samples_v3.20130502.ALL.panel/sample/super_pop=EUR

time ssimp --gwas EP.txt ~/reference_panels/1000genomes/ALL.chr1.phase3_shapeit2_mvncall_integrated_v5a.20130502.genotypes.vcf.gz --out SSIMP_EPILEPSY_EUR_chr1.txt --sample.names ~/reference_panels/1KG/integrated_call_samples_v3.20130502.ALL.panel/sample/super_pop=EUR

time ssimp --gwas CAD.txt ~/reference_panels/1000genomes/ALL.chr1.phase3_shapeit2_mvncall_integrated_v5a.20130502.genotypes.vcf.gz --out SSIMP_CAD_EUR_chr1.txt --sample.names ~/reference_panels/1KG/integrated_call_samples_v3.20130502.ALL.panel/sample/super_pop=EUR

time ssimp --gwas DLID.txt ~/reference_panels/1000genomes/ALL.chr1.phase3_shapeit2_mvncall_integrated_v5a.20130502.genotypes.vcf.gz --out SSIMP_DLID_EAS_chr1.txt --sample.names ~/reference_panels/1KG/integrated_call_samples_v3.20130502.ALL.panel/sample/super_pop=EAS

time ssimp --gwas CCANCER.txt ~/reference_panels/1000genomes/ALL.chr1.phase3_shapeit2_mvncall_integrated_v5a.20130502.genotypes.vcf.gz SSIMP_COL_CANCER_EAS_chr1.txt --sample.names ~/reference_panels/1KG/integrated_call_samples_v3.20130502.ALL.panel/sample/super_pop=EAS

time ssimp --gwas GFRAFR.txt ~/reference_panels/1000genomes/ALL.chr1.phase3_shapeit2_mvncall_integrated_v5a.20130502.genotypes.vcf.gz --out SSIMP_GFR_AFR_chr1.txt --sample.names ~/reference_panels/1KG/integrated_call_samples_v3.20130502.ALL.panel/sample/super_pop=AFR
``` 
 
##		DIST
Reference panel: 1000 Genomes Phase1 Release3 European https://drive.google.com/file/d/0By9SJLeBoduFTkM3NVBpbjB5ckU/view?resourcekey=0-q3TzfAsgU3IczIRZvfkAVQ
```{text}
time ./dist -c 1 ADHD.txt -o DIST_ADHD_EUR_chr1.txt -r 1kg.eur.ref.panel.dist/ref/chr1.1kg.eur.gz

time ./dist -c 1 UACR.txt -o DIST_UACR_chr1.txt -r 1kg.eur.ref.panel.dist/ref/chr1.1kg.eur.gz

time ./dist -c 1 GFREUR.txt -o DIST_GFR_EUR_chr1.txt -r 1kg.eur.ref.panel.dist/ref/chr1.1kg.eur.gz

time ./dist -c 1 EP.txt -o DIST_EPILPESY_EUR_chr1.txt -r 1kg.eur.ref.panel.dist/ref/chr1.1kg.eur.gz

time ./dist -c 1 CAD.txt -o DIST_CAD_EUR_chr1 -r 1kg.eur.ref.panel.dist/ref/chr1.1kg.eur.gz

time ./dist -c 1 DLID.txt -o DIST_D_LID_EAS_chr1.txt -r 1kg.eur.ref.panel.dist/ref/chr1.1kg.eur.gz

time ./dist -c 1 CCANCER.txt -o DIST_COL_CANCER_EAS_chr1.txt -r 1kg.eur.ref.panel.dist/ref/chr1.1kg.eur.gz

time ./dist -c 1 GFRAFR.txt -o DIST_GFR_AFR_chr1.txt -r 1kg.eur.ref.panel.dist/ref/chr1.1kg.eur.gz
``` 


 
##		FAPI
Reference panel: 1000 Genomes Project Phase 3, hg19 https://hgdownload.cse.ucsc.edu/gbdb/hg19/1000Genomes/phase3/ 
```{text}
time java -Xmx60000m -jar fapi.jar --impute --pfile ADHD.txt --gfile EUR_chr1.1kg.phase3.v5a.vcf.gz::vcf --no-web --out FAPI_ADHD.txt

time java -Xmx60000m -jar fapi.jar --impute --pfile UACR.txt --gfile EUR_chr1.1kg.phase3.v5a.vcf.gz::vcf --no-web --out FAPI_UACR.txt

time java -Xmx60000m -jar fapi.jar --impute --pfile GFREUR.txt --gfile EUR_chr1.1kg.phase3.v5a.vcf.gz::vcf --no-web --out FAPI_GFREUR.txt

time java -Xmx60000m -jar fapi.jar --impute --pfile EP.txt --gfile EUR_chr1.1kg.phase3.v5a.vcf.gz::vcf --no-web --out FAPI_EP.txt

time java -Xmx60000m -jar fapi.jar --impute --pfile CAD.txt --gfile EUR_chr1.1kg.phase3.v5a.vcf.gz::vcf --no-web --out FAPI_CAD.txt

time java -Xmx60000m -jar fapi.jar --impute --pfile DLID.txt --gfile EAS_chr1.1kg.phase3.v5a.vcf.gz::vcf --no-web --out FAPI_DLID.txt

time java -Xmx60000m -jar fapi.jar --impute --pfile CCANCER.txt --gfile EAS_chr1.1kg.phase3.v5a.vcf.gz::vcf --no-web --out FAPI_CCANCER.txt

time java -Xmx60000m -jar fapi.jar --impute --pfile GFRAFR.txt --gfile AFR_chr1.1kg.phase3.v5a.vcf.gz::vcf --no-web --out FAPI_GFRAFR.txt
```

 